### DISCUSSION TITLE
Support for CSS Modules

### DISCUSSION THREAD

**Discussion by chaance** [remix-run/remix#2214] [Proposals]
CSS Modules are a widely used method for dealing with scoped selectors in JavaScript projects, and in React they are often used as the basis for component-level styles. Many frameworks support them out-of-the-box, and Remix plans to do the same.

This discussion will be used to track the development of CSS Modules support as it is implemented. There are several pieces to the CSS Modules puzzle, and we'll likely roll out support for them in phases.

Another thing to consider is our approach in loading styles on the page. Many frameworks accomplish this by injecting a `<style>` tag into the head of the document as CSS modules are discovered (this may happen on the client or on the server). We prefer to generate static stylesheets, as static assets are easier to cache and this approach aligns more closely with our philosophy overall ([more on that here](https://github.com/evanw/esbuild/issues/20#issuecomment-802155704)). We could either generate a single stylesheet with all module styles defined in one place, or we could generate one stylesheet per route since we'll have the data for CSS Modules in the assets manifest. In the beginning we're likely to start with a global stylesheet as it simplifies both the implementation and the user API.

For reference: https://github.com/css-modules/css-modules

More to come, and questions/feedback welcome. 🚀

**Comment by dagda1** [comment]
I wonder if adding support for css modules could also open up support for vanilla-extract.  There is a [discussion about this](https://github.com/seek-oss/vanilla-extract/discussions/178) in vanilla-extract.  Vanilla extract does use a babel-plugin which is another problem.

**Reply by chaance** [reply]
Absolutely. We still hope to support it, and it's come up several times internally. It's on our roadmap after we get a few other styling priorities knocked out.

**Comment by didimedina** [comment]
Any updates on this? Really eager to try remixing vanilla-extract 🤣

**Reply by chaance** [reply]
No updates. You'll see it here when that changes.

**Comment by cherscarlett** [comment]
I would really, really like to move to remix, but this is a huge barrier for me. I'm considering developing a workaround, but would love to see this available out-of-the-box. I think adoption would soar from existing React projects that already utilize it.

**Reply by CanRau** [reply]
just in case, here's a workaround example https://github.com/jacob-ebey/remix-css-modules

**Reply by 0916dhkim** [reply]
@CanRau Can you briefly explain the workaround? I took a look at the repo, but I can't find where the workaround is.

**Reply by kiliman** [reply]
It's in *postcss.config.js*. After `postcss-module` transforms the CSS from _./styles/*.css_ to CSS module format, it generates a JSON file with the class names.
```ts
module.exports = {
  plugins: [
    require("postcss-modules")({
      getJSON: async (cssFilename, json, outputFilename) => {
        console.log(cssFilename, json, outputFilename);
        await fsp
          .mkdir(path.dirname(outputFilename), { recursive: true })
          .catch(() => {});
        await fsp.writeFile(
          `${outputFilename.replace(/\.css$/, ".json")}`,
          JSON.stringify(json)
        );
      },
    }),
  ],
};
```
```json
// package.json
"scripts: {
  "build:css": "postcss styles/**/*.css --dir app/styles --env production",
  "dev:css": "postcss styles/**/*.css --dir app/styles -w",
}
```
## CSS you write (outside _app_ folder)
```css
/* ./styles/index.css */
.container {
  color: darkslategrey;
  background-color: lightgrey;
}
```
## CSS and JSON that is generated (inside _app_ folder)
```css
/* generated file ./app/styles/index.css */
._container_162o7_7 {
  color: darkslategrey;
  background-color: lightgrey;
}
```
```json
// generated file ./app/styles/index.json
{"container":"_container_162o7_7"}
```
## Consume CSS Modules in your route
```ts
// routes/index.tsx
import stylesUrl from "~/styles/index.css"; // generated css
import styles from "~/styles/index.json";   // generated json
// export the stylesheet
export let links: LinksFunction = () => {
  return [{ rel: "stylesheet", href: stylesUrl }];
};
// use the generated classname from json file
export default function Index() {
  return <div className={styles.container}></div>
}
```

**Reply by 0916dhkim** [reply]
I see. Thank you so much for the explanation! That makes sense :100:

**Reply by dagda1** [reply]
isIs vanilla-extract now possible because it does not require babel now?

**Comment by convcha** [comment]
https://github.com/orgs/remix-run/projects/5/views/1
No CSS Modules support in roadmap😢

**Reply by machour** [reply]
Roadmap is not entirely published :)

**Reply by convcha** [reply]
I'm sorry, I was hasty😅

**Reply by chaance** [reply]
CSS Modules in the Remix compiler is still on hold for now. Keep your eye on https://github.com/orgs/remix-run/projects/5/views/1, we'll have some alternative approaches for you sooner than that!

**Reply by convcha** [reply]
Thank you!
I am very much looking forward to it😘

**Reply by jere-co** [reply]
Awesome. No Remix for me before Vanilla Extract is an option.

**Reply by machour** [reply]
@jere-co Remix for you then? https://github.com/remix-run/examples/tree/main/vanilla-extract

**Reply by jere-co** [reply]
@machour Cheers. That's the same approach that `markdalgleish` has in his `remix-vanilla-extract-prototype` - unfortunately quite serious tradeoffs there especially when the app grows. Hoping to see an official support soon, and confident it'll happen with the recent acquisitions.

**Reply by penx** [reply]
the acquisition of Mark? 😄

**Reply by sergiodxa** [reply]
> That's the same approach that markdalgleish has in his remix-vanilla-extract-prototype

That's because markdalgleish was the original author of the example

**Reply by markdalgleish** [reply]
@jere-co I'm keen to hear more from you about what an ideal Vanilla Extract + Remix integration would look like. As you said, that example I made is definitely _not_ that ideal.

**Comment by MichaelDeBoey** [comment]
Formalized in https://github.com/remix-run/remix/discussions/4542
